extension/ts/objects/bookForm.ts

```diff
@@ -0,0 +1,95 @@
+
+const getFormElements = (): Element[] => getElementsByTags(getForm(), FORM_DATA_ELEMENT_TAGS);
+
+// This is relying on the fact that when the edit form is available, the html matches this selector,
+// and fails to match in all other cases. This IS brittle. If LibraryThing changes
+// the markup in any way this will just not work properly
+const formExists = (): boolean => !!document.querySelector("#book_editForm > .book_bit");
+
+const getForm = (): HTMLElement => document.getElementById("book_editForm");
+
+const extractSaveDataFor = (targetElement: Element, saveData: FormData) => {
+	if (targetElement.id.startsWith(COLLECTIONS_ID_PREFIX)) {
+		const span = targetElement.parentElement.getElementsByTagName("span")[0];
+		return saveData[COLLECTIONS_KEY][span?.textContent] ?? targetElement;
+	} else {
+		return saveData[targetElement.id] ?? targetElement;
+	}
+};
+
+const getFormData = () => getFormElements()
+	.reduce((saveData: FormData, element: any) => {
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
+			const {value, checked} = element;
+			if (element.id.startsWith(COLLECTIONS_ID_PREFIX)) {
+				const collections = saveData[COLLECTIONS_KEY] || {};
+				const [span] = element.parentElement.getElementsByTagName("span");
+				collections[span.textContent] = {value, checked};
+				saveData[COLLECTIONS_KEY] = collections;
+			} else {
+				saveData[element.id] = {value, checked};
+			}
+		}
+		return saveData;
+	}, {});
+
+const insertFormData = (saveData: FormData) => getFormElements()
+	.forEach((element: any) => {
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
+			const {value, checked} = extractSaveDataFor(element, saveData);
+			if (element.value !== value || element.checked !== checked) {
+				element.value = value;
+				element.checked = checked;
+				element.dispatchEvent(new Event("change"));
+			}
+		}
+	});
+
+const forEachFormElement: ForEachFormElement = (callback: (element: Element) => void): void =>
+	getFormElements().forEach(callback);
+
+const onFormRender = (callback: (form: HTMLElement, forEachElement: ForEachFormElement) => void): void =>
+	window.addEventListener(FORM_RENDER_EVENT, () => callback(getForm(), forEachFormElement));
+
+/**
+ * I hate that this lives here,
+ * but I don't want to expose COLLECTIONS_KEY
+ *
+ * Also, this dumb implementation only works
+ * because the form is always the same order.
+ *
+ * This wouldn't work if the form could have elements rearranged
+ * @param formA
+ * @param formB
+ */
+const formDataEquals = (formA: FormData, formB: FormData): boolean =>
+	JSON.stringify(formA) === JSON.stringify(formB);
```

Comment:
AHAHAHAHAHA making this any smarter would be far too much effort and i doubt we'll run into any issues w/ it